### PR TITLE
scaletesting: bulkreposettings - allow repos to be incrementally fetched and updated

### LINKED ISSUE #44141: Scaletesting: Create 200k blank repos on GHE scaletesting
Due to the slowness we're experiencing on ghe.sgdev.org we've been forced to create the repos on ghe-scaletesting from scratch rather than copy the repos from ghe.sgdev.org.

**Organisation**: [blank200k](https://ghe-scaletesting.sgdev.org/blank200k)
- [x] Create 200k repos
- [x] Convert repos to private

### PR DESCRIPTION
**Pull request by burmudar**
Due to the large amount of repos on certain orgs, it takes a LONG time to fetch all of the repos before we start updating them.

This reworks how repos are fetched, by fetching a batch and then updating that batch.

Closes #44141 
## Test plan
Tested locally / remotely too (80k repos converted so far)

### CODE REVIEW

File: dev/scaletesting/bulkreposettings/main.go [hunk 1]
```diff
@@ -145,22 +186,117 @@ var app = &cli.App{
...
+		return nil
+	}
+
+	results, next, err := f.listRepos(ctx, f.org, f.pageStart, f.pageSize)
+	if err != nil {
+		f.err = err
+		return nil
+	}
+
+	hasMore := next != 0
```
Comment (burmudar) [inline dev/scaletesting/bulkreposettings/main.go:277]:
```suggestion
    // when next is 0, it means the Github api returned the nextPage as 0, which indicates that there are not more pages to fetch
	hasMore := next != 0
```

File: dev/scaletesting/bulkreposettings/main.go [hunk 2]
```diff
@@ -145,22 +186,118 @@ var app = &cli.App{
...
+		f.err = err
+		return nil
+	}
+
+    // when next is 0, it means the Github api returned the nextPage as 0, which indicates that there are not more pages to fetch
+	hasMore := next != 0
+	if !hasMore {
+		f.done = true
+	}
+
```
Comment (burmudar) [inline dev/scaletesting/bulkreposettings/main.go:282]:
```suggestion
// Ensure that the next request starts at the next page
```

File: dev/scaletesting/bulkreposettings/main.go [hunk 3]
```diff
@@ -86,57 +88,111 @@ var app = &cli.App{
...
+						}
+
+						if err := repoIter.Err(); err != nil {
+							logger.Error("repo iterator encountered an error", log.Error(err))
+						}
+
+						results := g.Wait()
+
+						// Check that we actually got errors
+						errs := make([]error, 0)
```
Comment (jhchabran) [inline dev/scaletesting/bulkreposettings/main.go:180]: nit simplify

```suggestion
						errs := []error{}
```

File: dev/scaletesting/bulkreposettings/main.go [hunk 4]
```diff
@@ -145,22 +201,118 @@ var app = &cli.App{
...
+}
+
+var _ Iter[[]*store.Repo] = (*GithubRepoFetcher)(nil)
+
+// StaticRepoFetcher satisfies the Iter interface allowing one to iterate over a static array of repos. To change
+// how many repos are returned per invocation of next, set iterSize (default 10). To start iterating at a different
+// index, set start to a different value.
+//
+// The iteration is considered done when start >= len(repos)
+type StaticRepoFetcher struct {
```
Comment (jhchabran) [inline dev/scaletesting/bulkreposettings/main.go:217]: nit: Usually, we're naming those things `Mock*` unless we're actually using them outside the tests.

```suggestion
type MockRepoFetcher struct {
```

File: dev/scaletesting/bulkreposettings/main.go [hunk 5]
```diff
@@ -145,22 +201,118 @@ var app = &cli.App{
...
+
+}
+
+type GithubRepoFetcher struct {
+	client    *github.Client
+	repoType  string
+	org       string
+	pageStart int
+	pageSize  int
+	done      bool
```
Comment (jhchabran) [inline dev/scaletesting/bulkreposettings/main.go:261]: nit: common naming

```suggestion
	page int
	perPage  int
	done      bool
```

File: dev/scaletesting/bulkreposettings/main.go [hunk 6]
```diff
@@ -145,22 +201,118 @@ var app = &cli.App{
...
+	err       error
+}
+
+// Done determines whether more repos can be retrieved from Github
```
Comment (jhchabran) [inline dev/scaletesting/bulkreposettings/main.go:265]:
```suggestion
// Done determines whether more repos can be retrieved from Github.
```

File: dev/scaletesting/bulkreposettings/main.go [hunk 7]
```diff
@@ -86,57 +88,111 @@ var app = &cli.App{
...
-									if err != nil {
-										r.Failed = err.Error()
-									} else {
-										r.Failed = ""
-										r.Pushed = true
-										break
-									}
-								}
+						for !repoIter.Done() && repoIter.Err() == nil {
+
```
Comment (jhchabran) [inline dev/scaletesting/bulkreposettings/main.go:128]:
```suggestion
```

**Review by jhchabran — APPROVED** [review]
LGTM! Mostly nits comments, don't wait for a second round if you decide to fix them.

**Pull request merged**